src/pynwb/base.py

```diff
@@ -282,3 +284,85 @@ def __init__(self, **kwargs):
+    response recordings, but a user can specify either only one of them or both. Since there is no
+    "None" value for a complex type like ``(idx_start, count, TimeSeries)``, we internally
+    define None as ``(-1, -1, TimeSeries)``, i.e., if the ``idx_start`` (and ``count``) is negative
+    then this indicates an invalid link (in practice both ``idx_start`` and ``count`` must always
+    either both be positive or both be negative). When selecting data via the
+    :py:meth:`~pynwb.base.TimeSeriesReferenceVectorData.get` or
+    :py:meth:`~pynwb.base. TimeSeriesReferenceVectorData.__getitem__`
+    functions, ``(-1, -1, TimeSeries)`` values are masked in the resulting np.ma.core.MaskedArray.
+    """
+    @docval({'name': 'name', 'type': str, 'doc': 'the name of this VectorData', 'default': 'timeseries'},
+            {'name': 'description', 'type': str, 'doc': 'a description for this column',
+             'default': "Column storing references to a TimeSeries (rows). For each TimeSeries this "
+                        "VectorData column stores the start_index and count to indicate the range in time "
+                        "to be selected as well as an object reference to the TimeSeries."},
+            *get_docval(VectorData.__init__, 'data'))
+    def __init__(self, **kwargs):
+        call_docval_func(super().__init__, kwargs)
+
+    def get(self, key, **kwargs):
+        """
+        Retrieve elements from this object.
+
+        The selection behavior is the same as for :py:meth:`~hdmf.common.table.VectorData.get`. A
+        key difference is that for invalid reference (i.e., values of ``(-1, -1, TimeSeries)``), the
+        value is being represented by a ``np.ma.core.MaskedArray`` instead. This allows us to
+        avoid exposing internal details of the schema to the user and simplifies handling of
+        missing values.
+
+        :param key: Selection of the elements
+        :param kwargs: Ignored
+        """
+        vals = super().get(key)
+        # we only selected one row.
+        if isinstance(key, (int, np.integer)):
+            # NOTE: If we never wrote the data to disk, then vals will be a single tuple.
+            #       If the data is loaded from an h5py.Dataset then vals will be a single
+            #       np.void object. I.e., an alternative check would be
+            #       if isinstance(vals, tuple) or isinstance(vals, np.void):
+            #          ...
+            if vals[0] < 0 or vals[1] < 0:
+                return np.ma.core.MaskedArray(data=vals, mask=[True, True, True])
+            else:
+                return vals
+        else:  # key selected multiple rows
+            # When loading from HDF5 we get an np.ndarray otherwise we get list-of-list. This
+            # makes things the values consistent
+            if isinstance(vals, np.ndarray):
+                vals = [[v[0], v[1], v[2]] for v in vals]
+            vals = np.asarray(vals)
```

Comment:
If vals is a numpy array, what does this sequence do?

Reply:
@rly I have to admit, this is a bit of a tricky problem. The case ``if isinstance(vals, np.ndarray)`` occurs when the data is being read from disk. In this case ``self.data`` will be a ``hdmf.backends.hdf5.h5_utils.ContainerH5TableDataset`` and the data is read from file as a numpy structured array. In contrast, when we create the file, the compound data is treated as a list of tuples. This means that the structure of the data is quite different in those  two cases.  So when calling ``np.asarray(vals)`` we get:
* A numpy array with ``dtype=object`` and ``shape=(1,3)`` when ``self.data`` is a list of tuples
* A numpy array with ``dtype=[('idx_start', '<i4'), ('count', '<i4'), ('timeseries', 'O')] `` and ``shape=(1,)`` when ``self.data`` is read from file. 

This leads to two key issues:

* Dealing with the difference in shape (i.e., 1D vs 2D array) downstream is annoying
* Dealing with the difference in dtype downstream is annoying
* Conversion to pandas DataFrame fails with the structured dtype with the error shown below.

In part because of the latter issue I had opted for converting the data on read to a list of tuples to keep things consistent. I agree that this should be documented more clearly in the code to avoid confusion later on.

How does VectorData usually deal with compound types? Are they treated as 2D arrays or as 1D structured arrays? Ideally we should try to keep the behavior here consistent with how other Container classes treat compound datatypes. If we should use a structured array instead (which may be more intuitive) it would require: 1) converting ``vals`` to a structured array if it is a list of tuples (essentially the opposite of the current logic) and 2) require dealing with the conversion to pandas issues downstream. I'm not sure yet what it will take to deal with the conversion issue. Pandas appears to attempt a reduction on the masked array which fails to deal with the complex dtype. 

What are your thoughts? What should the right behavior be?

```
======================================================================
ERROR: test_add_row_no_stimulus (test_icephys_metadata_tables.IntracellularRecordingsTableTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "[...]/pynwb/tests/unit/test_icephys_metadata_tables.py", line 316, in test_add_row_no_stimulus
    self.write_test_helper(ir)
  File “[…]/hdmf/src/hdmf/utils.py", line 583, in func_call
    return func(args[0], **pargs)
  File "[...]/pynwb/tests/unit/test_icephys_metadata_tables.py", line 125, in write_test_helper
    assert_frame_equal(ir.to_dataframe(**to_dataframe_kwargs), in_ir.to_dataframe(**to_dataframe_kwargs))
  File "[...]/hdmf/src/hdmf/utils.py", line 583, in func_call
    return func(args[0], **pargs)
  File "[...]/pynwb/src/pynwb/icephys.py", line 616, in to_dataframe
    res = super().to_dataframe(ignore_category_ids=getargs('ignore_category_ids', kwargs))
  File "[...]/hdmf/src/hdmf/utils.py", line 583, in func_call
    return func(args[0], **pargs)
  File "[...]/hdmf/src/hdmf/common/alignedtable.py", line 246, in to_dataframe
    dfs += [category.to_dataframe().reset_index() for category in self.category_tables.values()]
  File "[...]/hdmf/src/hdmf/common/alignedtable.py", line 246, in <listcomp>
    dfs += [category.to_dataframe().reset_index() for category in self.category_tables.values()]
  File "[...]/hdmf/src/hdmf/utils.py", line 583, in func_call
    return func(args[0], **pargs)
  File "[...]/hdmf/src/hdmf/common/table.py", line 1055, in to_dataframe
    ret = self.__get_selection_as_df(sel)
  File "[...]/hdmf/src/hdmf/common/table.py", line 966, in __get_selection_as_df
    ret = pd.DataFrame(df_input, index=pd.Index(name=self.id.name, data=id_index))
  File "[...]/anaconda3/envs/icephys-port/lib/python3.8/site-packages/pandas-1.3.0-py3.8-macosx-10.9-x86_64.egg/pandas/core/frame.py", line 614, in __init__
    mgr = dict_to_mgr(data, index, columns, dtype=dtype, copy=copy, typ=manager)
  File "[...]/anaconda3/envs/icephys-port/lib/python3.8/site-packages/pandas-1.3.0-py3.8-macosx-10.9-x86_64.egg/pandas/core/internals/construction.py", line 462, in dict_to_mgr
    return arrays_to_mgr(
  File "[...]/anaconda3/envs/icephys-port/lib/python3.8/site-packages/pandas-1.3.0-py3.8-macosx-10.9-x86_64.egg/pandas/core/internals/construction.py", line 122, in arrays_to_mgr
    arrays = _homogenize(arrays, index, dtype)
  File "[...]/anaconda3/envs/icephys-port/lib/python3.8/site-packages/pandas-1.3.0-py3.8-macosx-10.9-x86_64.egg/pandas/core/internals/construction.py", line 577, in _homogenize
    val = sanitize_array(
  File "[...]/anaconda3/envs/icephys-port/lib/python3.8/site-packages/pandas-1.3.0-py3.8-macosx-10.9-x86_64.egg/pandas/core/construction.py", line 497, in sanitize_array
    data = sanitize_masked_array(data)
  File "[...]/anaconda3/envs/icephys-port/lib/python3.8/site-packages/pandas-1.3.0-py3.8-macosx-10.9-x86_64.egg/pandas/core/construction.py", line 454, in sanitize_masked_array
    if mask.any():
  File "[...]/anaconda3/envs/icephys-port/lib/python3.8/site-packages/numpy/core/_methods.py", line 45, in _any
    return umr_any(a, axis, dtype, out, keepdims)
TypeError: cannot perform reduce with flexible type

```

Reply:
In my opinion, compound-data-type data should always be stored in memory as a numpy structured array. A tuple (or list of tuples) is hard to interpret without the field names. This would also help move us away from building functionality around lists and tuples. We could define `TimeSeriesReferenceVectorData.add_row` to take in a tuple and convert it to the correct structured array element (or take in a structured array object and check its fields) before appending it to the internal numpy structured array. 

And to answer your question about VectorData and compound types, see https://github.com/hdmf-dev/hdmf/issues/655. If the compound type is defined in the schema, then before write, I believe it is  retrieved as a list of tuples, and after write, it is retrieved as a 1D numpy structured array.

Reply:
I'm in favor of using structured array here. If we should use the structured array type, do you have any idea about the pandas conversion failure in that case?

Reply:
From the stack trace, it seems like the issue is with the fact that the structured array is masked. I'm digging into it.

Reply:
I believe this is a not-yet-resolved bug in numpy: https://github.com/numpy/numpy/issues/7492 
that is not addressed by pandas: https://github.com/pandas-dev/pandas/issues/42739 (note there is a concerning comment there about masked records being deprecated)

We can get around it by calling `.filled()` on the masked array before passing it to `pd.DataFrame(...)` in `DynamicTable.__get_selection_as_df` but then `IntracellularRecordingsTable.to_dataframe` needs to be updated because you will no longer get a masked array in the dataframe, but a normal array with a fillvalue for masked data.

Reply:
Numpy/pandas might be able to resolve this issue, but it would not help us unless we want to force users to use only the latest numpy/pandas version.

Reply:
See https://github.com/hdmf-dev/hdmf/pull/658 for a fix in `DynamicTable.__get_selection_as_df`

Reply:
I guess the question is whether we should look for an alternative to using numpy masked array here. I think masked arrays are a convenient way to describe missing data values, but if they cause too much trouble, then we could also opt for an alternative representation. E.g., if we always used list of tuples then we could use ``(None, None, None)`` for masked elements.

Reply:
If we end up filling the masked array anyways, then it seems it would probably make more sense to either do this here directly (to avoid inconsistency between what a user sees here and when converting to pandas) or to change the representation for missing values altogether. I'm honestly not sure what ``filled`` will use as fill value in this case.

Reply:
Based on the fact that pandas doesn't support the masked array behavior well we decided to instead use a format where the return is a list of numpy structured arrays/constants and None for missing values.

Reply:
The following commit https://github.com/NeurodataWithoutBorders/pynwb/pull/1349/commits/79271f90c8712cc2da4c5ae52bd91d5d0fad041c implements the changes we had discussed to use lists of ``namedtuple`` instead of numpy masked arrays.